scipy/sparse/_csparsetools.pyx.in

```diff
@@ -162,6 +162,57 @@ cpdef int lil_insert(cnp.npy_intp M, cnp.npy_intp N, object[:] rows,
         else:
             data[pos] = x
 
+def lil_get_lengths(object[:] input,
+                    cnp.ndarray output):
+    return _LIL_GET_LENGTHS_DISPATCH[output.dtype](input, output)
+
+{{for NAME, T in get_dispatch(IDX_TYPES)}}
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def _lil_get_lengths_{{NAME}}(object[:] input,
+                    cnp.ndarray[{{T}}] output):
+    for i in range(len(input)):
+        output[i] = len(input[i])
+
+{{endfor}}
+
+{{define_dispatch_map('_LIL_GET_LENGTHS_DISPATCH', '_lil_get_lengths', IDX_TYPES)}}
+
+def lil_get_sum_lengths(object[:] input,
+                    cnp.ndarray output):
+    return _LIL_GET_SUM_LENGTHS_DISPATCH[output.dtype](input, output)
+
+{{for NAME, T in get_dispatch(IDX_TYPES)}}
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def _lil_get_sum_lengths_{{NAME}}(object[:] input,
+                    cnp.ndarray[{{T}}] output):
+    cdef {{T}} z = 0
+    for i in range(len(input)):
+        z += len(input[i])
+        output[i] = z
+
+{{endfor}}
+
+{{define_dispatch_map('_LIL_GET_SUM_LENGTHS_DISPATCH', '_lil_get_sum_lengths', IDX_TYPES)}}
+
+def lil_flatten_to_array(object[:] input,
+                         cnp.ndarray output):
+    return _LIL_FLATTEN_TO_ARRAY_DISPATCH[output.dtype](input, output)
+
+{{for NAME, T in get_dispatch(VALUE_TYPES)}}
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def _lil_flatten_to_array_{{NAME}}(object[:] input not None, cnp.ndarray[{{T}}] output not None):
+    cdef size_t pos = 0
+    for i in range(len(input)):
+        for j in range(len(input[i])):
+            output[pos] = input[i][j]
```

Comment:
I suspect there's some tweaking possible here to reduce overhead in this hot loop. Have you tried looking at the annotated Cython output for these functions?

My main guess is that we could avoid a bunch of checking if we cdef a type for `input[i]`.

It might also be possible to write this inner loop as an assignment, which Cython should hopefully be able to compile down to a memcpy.

Reply:
Yep! Typing `input[i]` produced an additional 20% improvement relative to the current version of the pull requests. I'll include it in the next push. Sadly, since `input[i]` is a Python list and output is a Numpy ndarray, it cannot really be simplified to memcpy.